pkg/kubelet/cm/cpumanager/policy_options.go

```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpumanager
+
+import "fmt"
+
+const (
+	// RejectNonSMTAlignedOption is the name of the CPU Manager policy option
+	RejectNonSMTAlignedOption string = "reject-non-smt-aligned"
```

Comment:
I've been thinking about this for a while, and I think I like the name `full-cpus-only` for this option. It describes the intent appropriately, and it leaves it open for composability with other options that may build on this idea beyond just using it to decide to reject a pod with containers requesting an inappropriate number of cores to satisfy this constraint.

Reply:
I like the idea but was thinking that in Kubernetes context, the term CPU can translate to a virtual core or a physical core depending on whether hyperthreading is enabled on a node or not. Because of this `full-cpus-only` could be misinterpreted. How about `full-physical-core-only` or `full-core-only`?

Reply:
Hmm. That's true. Kubernetes does tend to use `cpu` as a general term that could mean either `cpu` or `core` in the traditional sense (i.e. where the hierarchy goes `sockets -> cpus -> cores`). Linux also doesn't play nicely with this naming hierarchy (at least not when it comes to looking in `/proc/cpuinfo`). 

With that in mind, the names `full-physical-core-only` or `full-core-only` also seem inappropriate because they could also lead to confusion (for example, the lowest level entity in my mental model is always a core, so talking about a "full" core is meaningless to me).

Reply:
Well, the terminology is a bit confusing which makes naming challenging but the reason I suggested `full-physical-core-only` or `full-core-only` because my understand is that the hierarchy is `sockets -> cores -> cpus`. In case of a system with the following configuration:
```
Thread(s) per core:   2 (HT enabled)
Core(s) per socket:   12
Socket(s):            4
```
it has 4 CPU sockets, each socket has 12 cores and each core can have two threads (which I was referring to as CPUs as they are the execution units).

Reply:
I guess that's where the confusion is though.

Much of the literature out there (and the way I've always referred to it in my 20+ years of working with this) follows the hierarchy `sockets -> cpus -> cores`. Just do a quick google search for "[cpus vs. cores](https://www.google.com/search?q=cores+vs.+cpus)" and see what comes up.

Unfortunately, the problem seems to be that these terms have always been somewhat ambiguous (and never really agreed upon between the OS and the architecture communities), leading to this confusion. In fact, I always forget that Linux actually swaps theses terms from how I think about them, as you pointed out above.

Doing a quick google search, it appears BSD is similar though:
```
FreeBSD/SMP: Multiprocessor System Detected: 24 CPUs
FreeBSD/SMP: 2 package(s) x 6 core(s) x 2 SMT threads
```

It could very well be that I'm the odd one out and just stuck in my old ways. I'm curious what others think though. Has there been a convergence around these terms contrary to my understanding?

Reply:
Great point! Sure, let's see what others think on this subject.

Reply:
My mental model was quite similar to @klueska 's. Problem is: it seems, like @swatisehgal pointed out, that kubernetes is naming thing in yet another slightly different way than OS community historically did and than HW manufacturers are doing.
In out case, I think is better to align as much as possible to k8s' naming, even though the naming choice k8s did are debatable in the first place (at least IMHO).

That said, going back to the `sockets -> cpus -> cores` model, while I share a similar background, in recent times we are having a rapidly evolving landscape on which new terms (or different meaning for the same terms) are more and more common, for example `packages` or `dies` or `chips` or even fancier concepts like `cpu complexes` (which is really what we could have called `cores` + `cache` + `i/o link` ?). Even naming the entity we refer as "virtual core" is challenging, should it a "virtual core", a "execution unit/port", a "hyperthread", or just a "core"? :)

Exactly because we have such a confusing and rapidly changing landscape I think it is unrealistic to have a really common and well defined terminlogy, so once again I think it's better to stick to the terms k8s already uses - and in general focus on our own community and norms -.

Reply:
in more concrete terms, I'm looking at https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/ and it seems that any possible solution will lead to ambiguity, because the docs and terms leave room for ambiguity. As much as I'd love to clarify, I think there is too much legacy to have a chance to improve in this area.

However the docs also hints at `vCPU`, which could be our best handle here.
My proposal is to rename as `full-pcpus-only` (note "p" for physical cpus) which is again admittedly suboptimal but at least hints at how cpumanager internally represents objects (which is also circling back to initial suggestion from @klueska ) and also hints at the vcpu vs pcpu split.

Reply:
Sounds reasonable to me!